File: csharp/sones-GraphMappingFramework/sones-GraphMappingFramework/DBObjects/VertexType/IVertexType.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using sones.GraphMappingFramework.DBObjects.Definitions;

namespace sones.GraphMappingFramework.DBObjects
{
    public interface IVertexType
    {
         Boolean IsSealed();
         Boolean HasParentType();
         Boolean HasChildTypes();
         Boolean IsAncestor(IVertexType myotherType);
         Boolean IsAncestor(IVertexType myotherType, Boolean myIncludeSelf);
         Boolean IsDescendant(IVertexType myotherType);
         Boolean IsDescendant(IVertexType myotherType, Boolean myIncludeSelf);
         List<IVertexType> GetDescendantTypes();
         List<IVertexType> GetDescendantTypes(Boolean myIncludeSelf);
         List<IVertexType> GetAncestorTypes();
         List<IVertexType> GetAncestorTypes(Boolean myIncludeSelf);
         List<IVertexType> GetKinsmenTypes();
         List<IVertexType> GetKinsmenTypes(Boolean myIncludeSelf);
         List<IVertexType> GetChildrenTypes();
         List<IVertexType> GetParentType();

        #region Attributes

         Boolean HasAttribute(String myAttributeName);
         Boolean HasAttributes(Boolean myIncludeAncestors);
         AttributeDefinition GetAttributeDefinition(String myAttributeName);
         List<AttributeDefinition> GetAttributeDefinitions(Boolean myIncludeAncestors);

        #endregion

        #region Properties

         Boolean HasProperty(String myPropertyName);
         Boolean HasProperties(Boolean myIncludeAncestors);
         PropertyDefinition GetPropertyDefinition(String myPropertyName);
         List<PropertyDefinition> GetPropertyDefinitions(Boolean myIncludeAncestors);
         List<PropertyDefinition> GetPropertyDefinitions(List<String> myPropertyNames);

        #endregion

        #region Incoming

         Boolean HasBinaryProperty(String myPropertyName);
         Boolean HasBinaryProperties(Boolean myIncludeAncestors);
         BinaryPropertyDefinition GetBinaryPropertyDefinition(String myPropertyName);
         List<BinaryPropertyDefinition> GetBinaryPropertyDefinitions(Boolean myIncludeAncestors);

        #endregion

        #region Incoming Edges

         Boolean HasIncomingEdge();
         Boolean HasIncomingEdges(Boolean myIncludeAncestors);
         IncomingEdgeDefinition GetIncomingEdgeDefinition(String myEdgeName);
         List<IncomingEdgeDefinition> GetIncomingEdgeDefinitions(Boolean myIncludeAncestors);

        #endregion

        #region Outgoing Edges

         Boolean HasOutgoingEdge(String myEdgeName);
         Boolean HasOutgoingEdges(Boolean myIncludeAncestors);
         OutgoingEdgeDefinition GetOutgoingEdgeDefinition(String myEdgeName);
         List<OutgoingEdgeDefinition> GetOutgoingEdgeDefinitions(Boolean myIncludeAncestors);

        #endregion

        #region Uniques

        Boolean HasUniqueDefinitions(Boolean myIncludeAncestors);
        List<UniqueDefinition> GetUniqueDefinitions(Boolean myIncludeAncestors);
        #endregion

        #region Indices
         Boolean HasIndexDefinitions(Boolean myIncludeAncestors);
         List<IndexDefinition> GetIndexDefinitions(Boolean myIncludeAncestors);

        #endregion
    }
}

File: java/sones-GraphMappingFramework/src/sones/GraphMapping/DBObjects/IVertexType.java
package sones.GraphMapping.DBObjects;

import java.util.List;

import sones.GraphMapping.Definitions.AttributeDefinition;
import sones.GraphMapping.Definitions.BinaryPropertyDefinition;
import sones.GraphMapping.Definitions.IncomingEdgeDefinition;
import sones.GraphMapping.Definitions.IndexDefinitions;
import sones.GraphMapping.Definitions.OutgoingEdgeDefinition;
import sones.GraphMapping.Definitions.PropertyDefinition;
import sones.GraphMapping.Definitions.UniqueDefinition;

public interface IVertexType {
	public boolean IsSealed();
	public boolean HasParentType();
	public boolean HasChildTypes();
	public boolean IsAncestor(IVertexType _otherType);
	public boolean IsAncestor(IVertexType _otherType, boolean _IncludeSelf);
	public boolean IsDescendant(IVertexType _otherType);
	public boolean IsDescendant(IVertexType _otherType, boolean _IncludeSelf);
	public List<IVertexType> GetDescendantTypes();
	public List<IVertexType> GetDescendantTypes(boolean _IncludeSelf);
	public List<IVertexType> GetAncestorTypes();
	public List<IVertexType> GetAncestorTypes(boolean _IncludeSelf);
	public List<IVertexType> GetKinsmenTypes();
	public List<IVertexType> GetKinsmenTypes(boolean _IncludeSelf);
	public List<IVertexType> GetChildrenTypes();
	public List<IVertexType> GetParentType();
	// Attributes
	public boolean HasAttribute(String _AttributeName);
	public boolean HasAttributes(boolean _IncludeAncestors);
	public AttributeDefinition GetAttributeDefinition(String _AttributeName);
	public List<AttributeDefinition> GetAttributeDefinitions(boolean _IncludeAncestors);
	// Properties
	public boolean HasProperty(String _PropertyName);
	public boolean HasProperties(boolean _IncludeAncestors);
	public PropertyDefinition GetPropertyDefinition(String _PropertyName);
	public List<PropertyDefinition> GetPropertyDefinitions(boolean _IncludeAncestors);
	public List<PropertyDefinition> GetPropertyDefinitions(List<String> _PropertyNames);
	// incoming
	public boolean HasBinaryProperty(String _PropertyName);
	public boolean HasBinaryProperties(boolean _IncludeAncestors);
	public BinaryPropertyDefinition GetBinaryPropertyDefinition(String _PropertyName);
	public List<BinaryPropertyDefinition> GetBinaryPropertyDefinitions(boolean _IncludeAncestors);
	// incoming edges
	public boolean HasIncomingEdge();
	public boolean HasIncomingEdges(boolean _IncludeAncestors);
	public IncomingEdgeDefinition GetIncomingEdgeDefinition(String _EdgeName);
	public List<IncomingEdgeDefinition> GetIncomingEdgeDefinitions(boolean _IncludeAncestors);
	// outgoing edges
	public boolean HasOutgoingEdge(String _EdgeName);
	public boolean HasOutgoingEdges(boolean _IncludeAncestors);
	public OutgoingEdgeDefinition GetOutgoingEdgeDefinition(String _EdgeName);
	public List<OutgoingEdgeDefinition> GetOutgoingEdgeDefinitions(boolean _IncludeAncestors);
	// uniques
	public boolean HasUniqueDefinitions(boolean _IncludeAncestors);
	public List<UniqueDefinition> GetUniqueDefinitions(boolean _IncludeAncestors);
	// indices
	public boolean HasIndexDefinitions(boolean _IncludeAncestors);
	public List<IndexDefinitions> GetIndexDefinitions(boolean _IncludeAncestors);
}
